strinject.c

```diff
@@ -0,0 +1,462 @@
+	off_t offset;
+
+	/* parse injection parameter */
+	if (parse_inject_param(STRINJECT_TYPE_FMT_STR, param, len, (void *) &count, &offset) != 0 &&
+		count == 0)
+		return 0;
+
+	/* allocate format string buffer */
+	fmt_str = (char *) real_malloc(2 * count + 1);
+	if (!fmt_str)
+		return 0;
+
+	for (i = 0; i < count; i++)
+		strncpy(&fmt_str[i * 2], "%s", 2);
+
+	/* set destination buffer */
+	p = (char *) real_malloc(len);
+	if (!p)
+		return 0;
+
+	if (offset > 0)
+		real_memcpy(p, buffer, offset);
+
+	real_memcpy(&p[offset], fmt_str, count * 2);
+	if (offset < len)
+		real_memcpy(p + offset + count * 2, buffer + offset + count * 2, len - offset - count * 2);
+
+	real_memcpy(buffer, p, len);
+
+	/* free buffer */
+	real_free(p);
+
+	/* free format string buffer */
+	real_free(fmt_str);
+
+	return 1;
+}
+
+/*
+ * inject overflow buffer
+ */
+
+int inject_buf_overflow(const char *param, void *buffer, size_t len)
+{
+	char *overflow_buf, *p;
+	int count, i;
+
+	off_t offset;
+
+	/* parse injection parameter */
+	if (parse_inject_param(STRINJECT_TYPE_BUF_OVERFLOW, param, len, (void *) &count, &offset) != 0 &&
+		count == 0)
+		return 0;
+
+	/* allocate format string buffer */
+	overflow_buf = (char *) real_malloc(count + 1);
+	if (!overflow_buf)
+		return 0;
+
+	for (i = 0; i < count; i++)
+		overflow_buf[i] = 0x41;
+
+	/* set destination buffer */
+	p = (char *) real_malloc(len);
+	if (!p) {
+		real_free(overflow_buf);
+		return 0;
+	}
+
+	if (offset > 0)
+		real_memcpy(p, buffer, offset);
+
+	real_memcpy(&p[offset], overflow_buf, count);
+	if (offset < len)
+		real_memcpy(p + offset + count, buffer + offset + count, len - offset - count);
+
+	real_memcpy(buffer, p, len);
+
+	/* free buffer */
+	real_free(p);
+
+	/* free format string buffer */
+	real_free(overflow_buf);
+
+	return 1;
+}
+
+/*
+ * inject file line
+ */
+
+int inject_file_line(const char *param, void *buffer, size_t len)
+{
+	char fpath[128], *p;
+
+	FILE *fp;
+	int line_count = 0, selected_line;
+
+	char line_buf[1024];
+	size_t inject_len = 0;
+
+	off_t offset;
+
+	/* parse injection parameter */
+	memset(fpath, 0, sizeof(fpath));
+	if (parse_inject_param(STRINJECT_TYPE_FILE_LINE, param, len, (void *) fpath, &offset) != 0)
+		return 0;
+
+	/* open file */
+	fp = real_fopen(fpath, "r");
+	if (!fp)
+		return 0;
+
+	/* get whole line count of file */
+	while (real_fgets(line_buf, sizeof(line_buf), fp) != NULL)
+		line_count++;
+
+	real_fseek(fp, 0, SEEK_SET);
+
+	/* get line buffer with randomely selected line number */
+	selected_line = rand() % line_count;
+
+	line_count = 0;
+	while (real_fgets(line_buf, sizeof(line_buf), fp) != NULL) {
+		if (line_count == selected_line) {
+			/* check whether line is empty */
+			if (real_strlen(line_buf) == 0)
+				break;
+
+			/* remove end line */
+			line_buf[real_strlen(line_buf) - 1] = '\0';
+
+			/* get injection length */
+			inject_len = real_strlen(line_buf) < (len - offset) ? real_strlen(line_buf) : (len - offset);
+
+			break;
+		}
```

Comment:
Some problems
- 379 will never be reached (fgets will return NULL if no chars and no \n)
- 382 will lose last char of last line if it isn't `\n` terminated
```
if (line_count == selected_line) {
    inject_len = real_strlen(line_buf);
    if (line_buf[inject_len - 1] == '\n')
        --inject_len;
    if (inject_len > len - offset)
        inject_len = len - offset;
    break;
}
```